[SelectionDAG] Support integer promotion for VP_STORE

lei137 opened the pull request:
Add integer promotion support for VP_STORE via legalization of extend and truncate of each form.

Patch commandeered from: https://reviews.llvm.org/D109377

lei137 committed:
[SelectionDAG] Support integer promotion for VP_STORE

Add integer promotion support for VP_STORE via
legalization of extend and truncate of each form.

Patch commandeered from: https://reviews.llvm.org/D109377

diggerlin commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -875,6 +878,23 @@ SDValue DAGTypeLegalizer::PromoteIntRes_LOAD(LoadSDNode *N) {
 
+SDValue DAGTypeLegalizer::PromoteIntRes_VP_LOAD(VPLoadSDNode *N) {
+  assert(!N->isIndexed() && "Indexed vp_load during type legalization!");
+  EVT NVT = TLI.getTypeToTransformTo(*DAG.getContext(), N->getValueType(0));
```
if the NVT is same as N->getValueType(0) , do we still need to create a new SDValue with the NVT ?

diggerlin replied:
do we want to add an assert here ?

topperc replied:
I don't see much point. It's fundamental to how the type legalizer works and we don't assert it anywhere else.

diggerlin commented:
it looks the patch is target impendent, I run the  `llc llvm/test/CodeGen/RISCV/rvv/vpload.ll` in powerpc, it hit a error 

LLVM ERROR: Invalid size request on a scalable vector.

topperc commented:
> it looks the patch is target independent, I run the `llc llvm/test/CodeGen/RISCV/rvv/vpload.ll -mtriple powerpc64-ibm-aix7.2.0.0` in powerpc, it hit a error
> 
> LLVM ERROR: Invalid size request on a scalable vector.

Scalable vectors are only supported for RISC-V vectors and AArch SVE. Any other target will crash. They would need to be scalarized before SelectionDAG by an IR pass which no one has written.

diggerlin commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -2216,6 +2242,50 @@ SDValue DAGTypeLegalizer::PromoteIntOp_STORE(StoreSDNode *N, unsigned OpNo){
+                             N->isCompressingStore());
+}
+
+SDValue DAGTypeLegalizer::PromoteIntOp_VP_LOAD(VPLoadSDNode *N, unsigned OpNo) {
```
I added some printf in the function.
it looks the test case llvm/test/CodeGen/RISCV/rvv/vpload.ll do not hit the function. 

is there any test case test the function?

topperc replied:
It's impossible to test. The mask for AArch64 and RISC-V never needs to be promoted. The VL operand is required to be legal when the node is constructed. SelectionDAGBuilder handles that by calling TLI.getVPExplicitVectorLengthTy().

This code can just be deleted.

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -2216,6 +2242,50 @@ SDValue DAGTypeLegalizer::PromoteIntOp_STORE(StoreSDNode *N, unsigned OpNo){
+  SDValue Operand = N->getOperand(OpNo);
+
+  if (OpNo >= 4) {
+    // The Mask or EVL. Update in place.
```
I suspect this is untested too.

lei137 replied:
This is actually needed or vpstore.ll and fixed-vectors-vpstore.ll
```
$ llvm-lit test/CodeGen/RISCV/rvv/fixed-vectors-vpstore.ll -v
-- Testing: 1 tests, 1 workers --
FAIL: LLVM :: CodeGen/RISCV/rvv/fixed-vectors-vpstore.ll (1 of 1)
******************** TEST 'LLVM :: CodeGen/RISCV/rvv/fixed-vectors-vpstore.ll' FAILED ********************
Exit Code: 2

Command Output (stderr):
--
RUN: at line 2: /home/lei/llvm/community/build/bin/llc -mtriple=riscv32 -mattr=+d,+zvfh,+v    -verify-machineinstrs < /home/lei/llvm/community/llvm-project/llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vpstore.ll | /home/lei/llvm/community/build/bin/FileCheck /home/lei/llvm/community/llvm-project/llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vpstore.ll
+ /home/lei/llvm/community/build/bin/llc -mtriple=riscv32 -mattr=+d,+zvfh,+v -verify-machineinstrs
+ /home/lei/llvm/community/build/bin/FileCheck /home/lei/llvm/community/llvm-project/llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vpstore.ll
PromoteIntegerOperand Op #1: t14: ch = vp_store<(store unknown-size into %ir.ptr, align 8)> t0, t12, t6, undef:i32, t9, t11

LLVM ERROR: Do not know how to promote this operator's operand!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
```

topperc replied:
Isn't that failing on Operand 1? I was saying Operand 4 is untested.

lei137 replied:
Sorry.  I thought you meant this function was also not needed.

lei137 committed:
update to use opaque pointer

lei137 commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -2216,6 +2242,50 @@ SDValue DAGTypeLegalizer::PromoteIntOp_STORE(StoreSDNode *N, unsigned OpNo){
+}
+
+SDValue DAGTypeLegalizer::PromoteIntOp_VP_LOAD(VPLoadSDNode *N, unsigned OpNo) {
+  assert(OpNo >= 3 && "Only know how to promote the mask or length!");
```
But this function is meant to be used to promote either mask for length so the assert need to be 3 no?  Since:
* If `OpNo == 3` promote mask.
* If `OpNo ==4` promote length.

lei137 committed:
remove PromoteIntOp_VP_LOA()

lei137 committed:
Mask and length never needs to be promoted since this is required to be legal
when the node is constructed.  SelectionDAGBuilder handles that by calling
TLI.getVPExplicitVectorLengthTy().

diggerlin reviewed (approved):
LGTM, but please wait a few days to see if Craig Topper has any further comments on the patch.
